h2/src/main/org/h2/command/dml/Insert.java

```diff
@@ -381,13 +394,17 @@ private boolean handleOnDuplicate(DbException de) {
 
         ArrayList<String> variableNames = new ArrayList<>(
                 duplicateKeyAssignmentMap.size());
-        Expression[] row = list.get(getCurrentRowNumber() - 1);
+        Expression[] row = (currentRow == null) ? list.get(getCurrentRowNumber() - 1) 
+        		: new Expression[columns.length];
```

Comment:
Use characters instead of spaces for indentation. Change code formatting settings in your IDE for this project to avoid such problems in the future. There is a sample configuration for Eclipse in `h2/src/installer/eclipse.settings` or you can configure it manually in any good enough IDE.